### ISSUE TITLE
The choice of API is problematic

### ISSUE THREAD

**Issue by aabghari** [mlpack/mlpack#2042]
Hi,

The choice of API to access the internal members of a class is problematic in mlpack. For example, in case of HMM, it's more efficient (both CPU and memory wise) if we keep the _initial_ and _transition_ matrix in log value. But this is not possible without breaking the existing APIs. If the API to access the _transition_ matrix were like this:

```
const arma::mat& GetTransition() { return transition;}
void SetTransition(const arma::mat &netTrans) {transition = newTrans;}

```
I would be easily  able to change the set API to something like this and change the code to use the 
_logTransition_ instead of _transition_ matrix without breaking the APIs:

```
const arma::mat& GetTransition() { return exp(logTransition);}

void SetTransition(const arma::mat &netTrans) {
    logTransition = log(newTrans);
}
```

but with the existing API, I cannot convert the _transition_ to the log version because the modifier returns a reference to _transition_ mat.

`arma::mat& Transition() { return transition; }`

So to make the change I need to break the API, remove the existing one and introduce a new one such as 

`arma::mat& LogTransition() { return logTransition; }`

to force the end user to change their code:

`hmm.Transition() = trans; --> hmm.LogTransitioin() = log(trans)`

or modify the existing API to:

`arma::mat& Transition() { return logTransition; }
`

and hope that the end user will change their code to get the correct result:

`hmm.Transition() = trans; --> hmm.Transitioin() = log(trans)`

I wonder what you recommend?

**Comment by zoq** [comment]
Opened https://github.com/mlpack/mlpack/pull/2043, checkout the test case for an example.

**Comment by aabghari** [comment]
Can you elaborate on your answer. I cannot see the solution.

**Comment by zoq** [comment]
Sorry, for the confusion I was talking about https://github.com/mlpack/mlpack/issues/2041. I will take a look at this issue later.

**Comment by rcurtin** [comment]
@aabghari I see what you mean and this is a little bit tricky to figure out how to do right.  Here's one idea:

We can add an extra `bool` template parameter to `HMM` and call it `UseLogProbabilities`; by default this is `false`, but if it's set to `true`, then `HMM` will internally store log probabilities in `transition` and `initial`.

In this way, it won't break reverse-compatibility for users who don't explicitly set `UseLogProbabilities` to true, and then when mlpack 4.0.0 comes around we can break reverse compatibility by changing the default for `UseLogProbabilities` to `true`.

This is only the start of an idea and there would be some additional thinking about how to do it right and nicely, but it would at least allow us to work natively in logspace for transition probabilities.  The internal implementation would also need to be adapted so that they depend on the value of `UseLogProbabilities`.

Let me know what you think or if you'd like me to spend a bit of time elaborating and clarifying the idea. :+1:

**Comment by aabghari** [comment]
I am not sure if adding a `bool` will fix the problem because the implementation for log space is different. The issue is not whether or not to store the `transition`  and  `initial` in log space per se. If `UseLogProbabilities` is true then the `transition` and `initial` are  in log space therefore the implementation is different for example all the multiplies becomes addition compare to the none log one. If I understand properly, we need two implementations and switch between them based on the `UseLogProbabilities`. I am not sure this is a right approach unless I am missing something. My issue is that this choice of modifier `arma::mat& Transition() { return transition; }` does not let to hide the implementation from the end user.

**Comment by rcurtin** [comment]
I agree, you are right that `arma::mat& Transition()` has no way to hide the underlying implementation detail.  But no matter what we do, we *have* to break reverse compatibility in mlpack 4.0.0 if we want to change this to work in logspace.

Suppose that we instead just created an additional member `logTransition`, and then used that in the computations (and just made sure that `transition` was equal to `log(transition)` when we finished training).  This seems like it would allow us to avoid breaking reverse compatibility, but if a user modifies `transition` by calling `Transition()` but does not correspondingly modify `LogTransition()`, then their HMM will not predict/compute with the modified transition matrix (since we use `logTransition` internally).

In essence, what I am saying is that there is no way to hide this from the end user, specifically because we have this `arma::mat& Transition() { return transition; }` function: when that is called, we have no way to know what was updated and if it is in sync with another version of the log transition matrix or something.

You are right though that the implementation would have to be different.  My suggestion would be this:

  1) Change the implementation so that during training, `transition` is always treated as though it is in logspace.
  2) If `UseLogProbabilities` is `false`, then at the end of training, take `transition = exp(transition)` (and now the transition probabilities are not in logspace).
  3) Create two implementations of `Estimate()`, `Generate()`, `Predict()` and `LogLikehood()` that change behavior based on whether `UseLogProbabilities` is `true` or `false`.

This is extra code for now, but when mlpack 4.0.0 comes around we can remove that extra code and have *only* the logspace implementation.

I'm open to other ideas, if you have them, but I don't currently see another way than what I just described.  Maybe you have a different idea that is cleaner? :)
